### ISSUE TITLE
H2Stream should be given factory and site in its constructor.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8909]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)| @Lukasa reported|
|-|-|
|Trac ID|trac#8909|
|Type|defect|
|Created|2016-11-18 14:11:13Z|
|Branch|https://github.com/twisted/twisted/tree/8909-lukasa-factoryandsite|

Encountered while adding HTTP/2 support to Daphne, the channels web server.

The HTTP Request object expects to be able to grab the 'site' and 'factory' properties from its parent channel. However, some *subclasses* of the Request object will want to be able to get those during the __init__ call (Daphne's does).

This is fine for HTTP/1.1 because the HTTPChannel is completely constructed when it creates the Request object. But in HTTP/2 that's not true: the H2Stream creates the HTTPChannel in its own __init__ method.

There are two possible fixes: we can either pass H2Stream the site and factory directly on its constructor, or we can delay the Request initialization. I think it'll be easier to do the first: they're formally part of the object so they should be created at object initialization time rather than being patched on as they are right now.

<details><summary>Searchable metadata</summary>

```
trac-id__8909 8909
type__defect defect
reporter__Lukasa Lukasa
priority__normal normal
milestone__None None
branch__8909_lukasa_factoryandsite 8909-lukasa-factoryandsite
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__None None
time__1479478273660554 1479478273660554
changetime__1480269265145422 1480269265145422
version__None None
owner__Lukasa Lukasa

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_1"></a>|@Lukasa commented|
|-|-|

In changeset a0574d9d6597f3c2927ea3bbb076ac99c461ed4a
```
#!CommitTicketReference repository="" revision="a0574d9d6597f3c2927ea3bbb076ac99c461ed4a"
Failing test for #8909
```

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_2"></a>|@Lukasa commented|
|-|-|

In changeset 0653dd005c64d298d1d4b9119e348b0d8e2c5e42
```
#!CommitTicketReference repository="" revision="0653dd005c64d298d1d4b9119e348b0d8e2c5e42"
Add topfile for #8909.
```

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_3"></a>|@Lukasa commented|
|-|-|

See the patch in: https://github.com/twisted/twisted/pull/597

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban set owner to @Lukasa|
|-|-|

Hi,

Thanks for the patch.

It looks good.

The test contains a bit of noise for my taste ... I had to re-read the test to follow up the factory value... which is overwritten in validateFactoryAndSite

The release notes fragment talks about a private API twisted.web._http2.H2Stream . Is that OK? As a Twisted users should I care about twisted.web._http2 ?

only minor comments,

Please check my comments and merge... assuming that you will fix the txchecker builder

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_5"></a>|@Lukasa commented|
|-|-|

In changeset 696ab9048e6bd3563a0ae61da3d5a5390ff78c0b
```
#!CommitTicketReference repository="" revision="696ab9048e6bd3563a0ae61da3d5a5390ff78c0b"
Failing test for #8909
```

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_6"></a>|@Lukasa commented|
|-|-|

In changeset dbc2f496f9f34260b7a99ecf80063c4f5ded22c0
```
#!CommitTicketReference repository="" revision="dbc2f496f9f34260b7a99ecf80063c4f5ded22c0"
Add topfile for #8909.
```

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_7"></a>|@Lukasa removed owner|
|-|-|

Ok, I've gone through and added the test helpers. I'm not sure about the private API, but honestly this only affects people writing custom Request subclasses and they *might* care.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban set owner to @Lukasa|
|-|-|

Changes look good. Thanks!

If people might need/want to customize this class, then you might want to expose this as a public API.

Is a bit strange to talk in the release notes about some fancy functionality which end users are not supposed to use it.

So, if this is still private API, I think that the release notes fragments should be a .misc

Happy to see the removal of variable names like `f, a, b`.

Please merge at will as I don't want to block this :)

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_9"></a>|@Lukasa set status to `closed`|
|-|-|

In changeset 0165a8372f830a59429a8cae0a189c5b74a26f9b
```
#!CommitTicketReference repository="" revision="0165a8372f830a59429a8cae0a189c5b74a26f9b"
Merge 8909-lukasa-factoryandsite: Pass the factory and site to the H2Stream constructor.

Author: lukasa
Reviewer: adiroiban
Fixes: #8909
```
